ls1intum/Artemis
# `Lectures`: Fix the order of lecture units when merging pdf files

### Checklist
#### General
#### Server

### Motivation and Context
When using the merge pdf feature as a student, the created pdf file contained all slides in the order of the upload of their respective units. This is an issue e.g. when last minute changes are necessary, sicnce recreating a unit breaks the order of the merged file.

Testing the merge feature as a editor is also quite hard right now. The merged pdf only contains already public lecture units. This should be changed for tutors / editors, their file should contain all 
### Description
When merging the files, they are now sorted by the order-attribute.

Aditionally, we now use `isAllowedToSeeLectureUnit()`. This method checks the role of the user, allowing tutors and above to create a merged pdf including hidden units. With the current design of the file ressource, we cannot use the usual aproach of calling this method with `null` and loading the user automatically. Instead, we have to parse the username from the token.

### Steps for Testing
Prerequisites:
- 1 Instructor, 1 Student, 1 Tutor

1. Create two new lecture units with pdf attatchments:
2. As a student, click the "merge pdfs" button
3. Change the order of the units as instructor
4. Re-Download the merged pdfs again
5. Ensure that in both downloads the order was used, meaning the second download should have the files fliped.
6. Add another unit with a release date in the future
7. Merge the PDF as student and tutor. The students' pdf should not contain the new unit, but it should be included into the result for the tutor.

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
